Handle permissions for BackYourStack subscribers

flickz opened the pull request:
Resolves https://github.com/opencollective/opencollective/issues/2622

The following checks are made when a user clicks the 'Back My Stack' button:

- Authentication, the user must be signed in.

- If the github profile about to be saved is an organization, ensure the logged in user is an active admin of the organization

- If the github profile is a personal account, ensure the logged in user profile matches the profile about to be saved.

### How to display errors:

I couldn't decide the best way to display error message regarding permission check on `/profile` page so temporarily  I'm using browser alert as shown below. 

<img width="688" alt="Screenshot 2019-11-13 at 4 20 07 AM" src="https://user-images.githubusercontent.com/15707013/68731243-ae5e3480-05cf-11ea-9f6e-7b17c65ed04d.png">

Not the best way I will love an input from the design team regarding this.

znarf commented on src/lib/github.js:
```diff
@@ -122,6 +122,24 @@ async function fetchProfile(login, accessToken) {
   return null;
 }
 
+async function fetchOrgMemebership(orgName, login, accessToken) {
```
Typo in fetchOrgMemebership

znarf commented:
@flickz have you changed it since I gave feedback?

flickz commented:
Yes.

flickz commented:
The checks in this PR ensure everything in the spec but I'm not clear on this:

> Allow update for any profile if the connected GitHub account match the stored GitHub account

What do you mean by "stored GitHub account"?

znarf commented:
@flickz sorry but, the spec is saying to be relaxed on "saving" but strict on "updating".

> we should implement a simple permission system to prevent someone **to edit** the BackYourStack profile of someone else once the subscription is setup.

> By default don't allow **to update** a saved profile

znarf commented:
> What do you mean by "stored GitHub account"?

From the spec:

> As part of this effort, we should save in dependencies.json the connected GitHub account when the transaction was triggered (if any)

flickz commented:
<img width="1440" alt="Screenshot 2019-11-18 at 10 42 37 PM" src="https://user-images.githubusercontent.com/15707013/69123622-27fb9400-0aa2-11ea-88c4-a523b8084e4f.png">

znarf commented on src/server/index.js:
```diff
@@ -231,10 +238,49 @@ nextApp.prepare().then(() => {
+          return res
+            .status(401)
+            .send(
+              'You have to be an active admin of the organization update organization profile.',
```
Check the copy

znarf commented on src/pages/profile.js:
```diff
@@ -118,6 +120,13 @@ export default class Profile extends React.Component {
   handleBackMyStack = async () => {
+    if (!this.props.loggedInUser) {
+      this.setState({
+        error: 'You need to sign in to back your stack',
```
Check the copy: Back Your Stack

znarf commented on src/server/index.js:
```diff
@@ -231,10 +238,49 @@ nextApp.prepare().then(() => {
+    if (
+      dependenciesFile &&
+      profileOrder &&
+      profileOrder.triggeredBy !== loggedInUsername
```
Where is that coming from?

znarf replied:
@flickz Why was that resolved?

flickz replied:
Should be a mistake but `profileOrder.triggeredBy !== loggedInUsername` shouldn't be there again since we're no longer enforcing user to sign in before backing their stack. 

`dependenciesFile && profileOrder ` are checks to confirm the profile has been saved before and user is about to update the profile.

znarf commented on src/server/index.js:
```diff
@@ -249,11 +295,7 @@ nextApp.prepare().then(() => {
-    const savedId = await saveProfile(
-      data.profile.login,
-      repositories,
-      excludedRepos,
```
Why remove excludedRepos?

flickz replied:
I removed it because we don't need it anymore, once the  selected repositories are filtered only their files and keys will be saved in `dependecies.json`  making it easy to retrieve at once without further filtering.

znarf replied:
But that is not part of the permissions right?

flickz replied:
Yes.

flickz commented:
> I'm not sure how to test this now that the alternative banner is displayed. I'm curious about your approach.

Just for testing purpose, I usually comment out the alternative banner code in profile.js in order to test or I delete the profile on s3.

flickz commented:
> Please address the triggeredBy situation

The goal is to "Allow update for any profile if the connected GitHub account match the stored GitHub account",  the only way to store connected GitHub account is if the user is authenticated during "Back my stack", it will be saved as `triggeredBy`  during the dispatch process. I removed `triggeredBy`  because authentication was no longer compulsory but today, I remembered we concluded we still want it stored in the case where the user is authenticated, I've opened this [PR](https://github.com/opencollective/backyourstack/pull/281) to take care of storing `triggeredBy` in `profileOrder` when dispatch is about to be triggered and I've also updated the permission condition back. 

I think we're good to go now.
